brouter-core/src/main/java/btools/router/RoutingContext.java

```diff
@@ -286,14 +290,13 @@ public void unsetWaypoint()
 
   public int calcDistance( int lon1, int lat1, int lon2, int lat2 )
   {
-    double l = (lat2 - 90000000) * 0.00000001234134;
-    double l2 = l*l;
-    double l4 = l2*l2;
-    coslat = 1.- l2 + l4 / 6.;
-    double coslat6 = coslat*0.000001;
+    CheapRulerSingleton cr = CheapRulerSingleton.getInstance();
+
+    coslat = cr.cosIlat(lat2);
+    double coslat6 = coslat*cr.ILATLNG_TO_LATLNG;

```

Comment:
I kept the faster approximate distance computation below, and only compute the true distance at the end of the loop.

However, this code is very similar to the one found in  `WaypointMatcherImpl` and I think this could be factored into a common function call. What do you think?

Reply:
Isn't what CheapRuler.distance(..) uses internally as "kx", "ky" more or less the same then "coslat6", "0.00001" ? Except for that facrtor 111894.

So for me it seems you should'nt provide a coslat-lookup, but a kx,ky lookup, because that's basically what's needed all over the place. So by replacing coslat6 + 0.00001 -> kx+ky we could consistently redefine all these internal radius/distance measurements that are now in "cosinius normalized degrees" to be measures in meters, and everything woul be much more readable.

It would still be ugly code. RoutingContext.calcDistance does:
- distance calculation
- nogo matching
- start/endpoint matching
- cosinus calculation for later use

so it's result  is not just the return-value, but also rc.coslat, rc.shortestmatch, rc.wayfraction, nogo.radius - this is really spaghetti, but refactoring that maybe you should do some warmup first :-)

And youare right, WaypointMatcherImpl is doing similar things, because it's solving the same problem: finding the closest match of a point to a line connecting to other points.

Reply:
> Isn't what CheapRuler.distance(..) uses internally as "kx", "ky" more or less the same then "coslat6", "0.00001" ? Except for that facrtor 111894.

Indeed, that's exactly it. I was trying to avoid recomputing `kx` and `ky` since it involves computing powers of `cos` terms. But it could totally be cached and having a lookup for this would probably make sense. I'll go for it :)

> this is really spaghetti, but refactoring that maybe you should do some warmup first :-)

I'll try to write something to address this, try to refactor a bit the code and fix the remaining points. Feel free to merge this PR if you think it is ok. I can either push additional stuff in this one or open a new one, as it suits you best.